File: minimum-balance-ts/src/agent.ts
import BigNumber from 'bignumber.js'
import { 
  BlockEvent, 
  Finding, 
  HandleBlock, 
  FindingSeverity, 
  FindingType,
  getEthersProvider,
  ethers
} from 'forta-agent'

export const ACCOUNT = "0x6efef34e81fd201edf18c7902948168e9ebb88ae"
export const MIN_BALANCE = "500000000000000000" // 0.5 eth

const ethersProvider = getEthersProvider()

function provideHandleBlock(ethersProvider: ethers.providers.JsonRpcProvider): HandleBlock {
  return async function handleBlock(blockEvent: BlockEvent) {
    // report finding if specified account balance falls below threshold
    const findings: Finding[] = []

    const accountBalance = new BigNumber((await ethersProvider.getBalance(ACCOUNT, blockEvent.blockNumber)).toString())
    if (accountBalance.isGreaterThanOrEqualTo(MIN_BALANCE)) return findings

    findings.push(
      Finding.fromObject({
        name: "Minimum Account Balance",
        description: `Account balance (${accountBalance.toString()}) below threshold (${MIN_BALANCE})`,
        alertId: "FORTA-6",
        severity: FindingSeverity.Info,
        type: FindingType.Suspicious,
        metadata: {
          balance: accountBalance.toString()
        }
      }
    ))

    return findings
  }
}


export default {
  provideHandleBlock,
  handleBlock: provideHandleBlock(ethersProvider)
}

File: minimum-balance-py/src/agent.py
from forta_agent import Finding, FindingType, FindingSeverity, get_web3_provider, Web3

ACCOUNT = "0x6efef34e81fd201edf18c7902948168e9ebb88ae"
MIN_BALANCE = int("500000000000000000")  # 0.5 eth


def provide_handle_block(w3):
    def handle_block(block_event):
        findings = []

        print(f'block_number={block_event.block_number}')
        balance = int(w3.eth.get_balance(
            Web3.toChecksumAddress(ACCOUNT), int(block_event.block_number)))
        if (balance >= MIN_BALANCE):
            return findings

        findings.append(Finding({
            'name': "Minimum Account Balance",
            'description': f'Account balance ({balance}) below threshold ({MIN_BALANCE})',
            'alert_id': "FORTA-6",
            'severity': FindingSeverity.Info,
            'type': FindingType.Suspicious,
            'metadata': {
                'account': ACCOUNT,
                'balance': balance
            }
        }))
        return findings

    return handle_block


w3 = get_web3_provider()
real_handle_block = provide_handle_block(w3)